Memory allocation refactor

Suthiro opened the pull request:
I decided to do a small refactor in `Fit.cpp` and its friends to get rid of the manual memory management. Everything looks fine, but I'd like to get a code review. Also, I would like to discuss two specific parts of code (see in code comments).

Suthiro committed:
Memory allocation refactor

Suthiro commented on libscidavis/src/Convolution.cpp:
```diff
@@ -93,32 +88,31 @@ void Convolution::setDataFromTable(Table *t, const QString &signalColName,
     }
 
-    d_n = rows;
-
     d_n_signal = 16; // tmp number of points
-    while (unsigned(d_n_signal) < d_n + d_n_response / 2)
+    while (unsigned(d_n_signal) < rows + d_n_response / 2)
         d_n_signal *= 2;
 
-    d_x = new double[d_n_signal]; // signal
-    d_y = new double[d_n_response]; // response
-
-    if (d_y && d_x) {
-        memset(d_x, 0, d_n_signal * sizeof(double)); // zero-pad signal data array
-        for (unsigned i = 0; i < d_n; i++)
-            d_x[i] = d_table->cell(i, signal_col);
-        for (int i = 0; i < d_n_response; i++)
-            d_y[i] = d_table->cell(i, response_col);
-    } else {
```
`d_n` is supposed to reflect the size of data created with `new`. This is the only time in the whole project when d_n is assigned arbitrarily, because `d_x` is created with size `d_n_signal` and `d_y` with `d_n_response`. However, iteration for `d_x` goes over `d_n`. Is it bug or I don't understand something?

highperformancecoder replied:
No that is correct. d_n_signal is the power of two that is just greater of equal to the number of rows. d_x is filled with zeros before starting, and then rows data is copied into it.

highperformancecoder replied:
The way I would do it is make d_x and d_y vectors. You can do d_x=move(vector<double>(d_n_signal)). For d_y, better to use clear, reserve and push_back to avoid an unnecessary zero initialisation. You can get rid of d_n_signal and d_n_response as member variables, since they are just the sizes of d_x and d_y.

Suthiro replied:
>No that is correct. d_n_signal is the power of two that is just greater of equal to the number of rows. d_x is filled with zeros before starting, and then rows data is copied into it.

No doubt. But the iteration goes over `d_n` which is not necessarily equal to `d_n_signal`. Some rows are left uncopied from `signal_col` to `d_x`.

>The way I would do it is make d_x and d_y vectors. You can do d_x=move(vector(d_n_signal)). For d_y, better to use clear, reserve and push_back to avoid an unnecessary zero initialisation. 

`d_x` and `d_y` are made to be vectors in this PR. Actually, all of this hassle is to make 'em `std::vector`s (see `Filter.h` lines 136-139). Also next lines resize and fill vectors with zeroes. No point using `push_back` when the size is known.

>You can get rid of d_n_signal and d_n_response as member variables, since they are just the sizes of d_x and d_y.

Then I'm to remove `d_n` (or `d_n`) also, because what does it represent then? Size of what?

highperformancecoder replied:
d_n represents the number of rows copied into the data. At the end of the routine, that many items are copied back out.

Think FFT, where the input needs to be a power of 2. If the input isn't, then this algorithm zero pads to the next power of two.

resize() only zeros out elements added to the vector. clear() followed by resize() zeros the entire vector. Actually, strictly speaking, only the last d_n_signal-d_n  need to be zeroed.

Unnecessary zeroing of vectors can be a source of performance drag - which is why reserve() (which allocates the memory, but doesn't zero it), followed by push_back of the data is a good choice. In d_x's case, you can do a resize after the data is copied in, which zeros out the remaining elements.

In any case, SciDAVis is not exactly a high performance application, so it really doesn't matter much. I just like to "do things right" :).

highperformancecoder replied:
Sorry - it appears I didn't fully review this PR. I've run out of time today - I'll swing back on this tomorrow.

Suthiro replied:
> d_n represents the number of rows copied into the data. At the end of the routine, that many items are copied back out.
>
> Think FFT, where the input needs to be a power of 2. If the input isn't, then this algorithm zero pads to the next power of two.

You're right, I'll fix it.

> resize() only zeros out elements added to the vector. clear() followed by resize() zeros the entire vector. Actually, strictly speaking, only the last d_n_signal-d_n need to be zeroed.
> 
> Unnecessary zeroing of vectors can be a source of performance drag - which is why reserve() (which allocates the memory, but doesn't zero it), followed by push_back of the data is a good choice. In d_x's case, you can do a resize after the data is copied in, which zeros out the remaining elements.
> 
> In any case, SciDAVis is not exactly a high performance application, so it really doesn't matter much. I just like to "do things right" :).

Right now I'm removing all `d_n()` calls from the code since and replacing with appropriate `std::vector::size()`. Also, I will rewrite the initializers to `resize()` and `push_back()` where possible.

Suthiro commented on libscidavis/src/Fit.cpp:
```diff
@@ -603,14 +602,8 @@ void Fit::generateFitCurve(const vector<double> &par)
     if (d_gen_function) {
         insertFitFunctionCurve(objectName() + tr("Fit"), X, Y);
         d_graph->replot();
-    } else {
-        auto x = new double[d_points];
-        std::copy(X.cbegin(), X.cend(), x);
-        auto y = new double[d_points];
-        std::copy(Y.cbegin(), Y.cend(), y);
-        // according to addFitCurve docstring "..and frees the input data from memory."
-        d_graph->addFitCurve(addResultCurve(x, y));
-    }
+    } else
+        d_graph->addFitCurve(addResultCurve(X, Y));
```
Another interesting story. There was a comment "according to addFitCurve docstring ..and frees the input data from memory." But I'm almost sure that it is does not. `addResultCurve` has two overloads, and both of them do `memcpy`, leaving the input data intact.
`addFitCurve` is implemented as
`void Graph::addFitCurve(QwtPlotCurve *c)
{
    if (c)
        d_fit_curves << c;
}`
So the curve is placed into a list. I suppose that the data inside `QwtPlotCurve` will be freed upon destruction, but it is not the same data (memory-wise) that passed to `addResultCurve`.
I think that it was a memory leak, but again, I'd like to get another opinion here. Thanks.

highperformancecoder replied:
It took a bit of drilling into the code, including looking up the Qwt5 setData call, but I do believe you're right. Let's lock it in.

highperformancecoder commented on libscidavis/src/Convolution.cpp:
```diff
@@ -93,32 +88,31 @@ void Convolution::setDataFromTable(Table *t, const QString &signalColName,
+        return;
     }
+    std::fill(d_x.begin(), d_x.end(), 0.0);
+    std::fill(d_y.begin(), d_y.end(), 0.0);
```
This fill is unnecessary

highperformancecoder commented on libscidavis/src/Convolution.cpp:
```diff
@@ -93,32 +88,31 @@ void Convolution::setDataFromTable(Table *t, const QString &signalColName,
     }
+    std::fill(d_x.begin(), d_x.end(), 0.0);
+    std::fill(d_y.begin(), d_y.end(), 0.0);
+    for (unsigned i = 0; i < d_n_signal; i++)
```
This loop bound should be rows, as d_n_signal may be too large for the following cell() call.

highperformancecoder commented on libscidavis/src/FFT.cpp:
```diff
@@ -190,27 +192,23 @@ void FFT::setDataFromTable(Table *t, const QString &realColName, const QString &
+        return;
+    }
+
+    std::fill(d_y.begin(), d_y.end(), 0.0);
```
If you clear d_y before resizing, that would gurantee it is fully zero-filled

highperformancecoder commented on libscidavis/src/Filter.h:
```diff
@@ -128,13 +130,13 @@ class Filter : public QObject
 
     //! Size of the data arrays
-    unsigned d_n;
+    size_t d_n() const { return d_x.size(); };
```
Is it always this? I thought the example of FFT meant that sometimes d_n was less than d_x.size().

Suthiro replied:
This was an outdated revision, I've removed `d_n()` everywhere because of this in 0d092db.

highperformancecoder commented:
Let not worry too much about optimum use of reserve, resize, push_back etc. However, I do think we need to be careful on the use of d_n as being the number of rows of the data column, vs the size of the vector being filtered (which in the case of FFTs must be a power of 2).

Maybe try renaming d_n to numRows and numTransformed to keep it straight what the usage is?

Suthiro commented:
> Let not worry too much about optimum use of reserve, resize, push_back etc. However, I do think we need to be careful on the use of d_n as being the number of rows of the data column, vs the size of the vector being filtered (which in the case of FFTs must be a power of 2).
> 
> Maybe try renaming d_n to numRows and numTransformed to keep it straight what the usage is?

You've meant `d_n` to `numRows` and `d_n_signal` to `numTransformed`?

Suthiro committed:
PR review-based changes & bugfix for SF#214

Suthiro commented:
I've added changes to the code. What has been changed:
I fixed zero-padding in FFT and re-checked the changes. I removed some now-useless member variables and replaced them with appropriate `std::vector::size` calls. While `std::vector::reserve` + `std::vector::push_back` looks good, it does not shrinks the vector, so either `std::vector::shrink_to_fit` or `std::vector::resize` must be called in case when new size is smaller than the current one. So I decided to stick with `std::vector::resize` + `std::fill(not-the-begin-iterator,end-iterator,0.0)` to avoid unnecessary zeroing. Also, I've found the source of [SF#214 bug](https://sourceforge.net/p/scidavis/scidavis-bugs/214/) and fixed it. In short: when the x-y data was sorted, y_errors were not. I believe there are more related tickets on SF.
I see that some are not passing, so I will take a look on 'em tomorrow.

Suthiro committed:
FFT segfault fix

Suthiro commented:
Looks like everything is okay now.

highperformancecoder commented on libscidavis/src/Filter.h:
```diff
@@ -170,4 +165,30 @@ class Filter : public QObject
+        std::vector<std::pair<T,size_t>> tData(data.size());
+        size_t ii=0;
+        std::generate(tData.begin(), tData.end(), [&](){ auto pair = std::make_pair(data[ii],ii); ++ii; return pair; });
+        std::sort(tData.begin(),tData.end(),[](const std::pair<T,size_t>& a, std::pair<T,size_t>& b){ return a.second < b.second;});
```
This seems to be a null operation. The second field of pair are the integers 0..data.size()-1.

Suthiro replied:
Yes, you're right. I wonder why it was working. I changed the lines to sort on `.first` and to use `std::bound` methods.

highperformancecoder commented on libscidavis/src/Filter.h:
```diff
@@ -170,4 +165,30 @@ class Filter : public QObject
+        std::generate(tData.begin(), tData.end(), [&](){ auto pair = std::make_pair(data[ii],ii); ++ii; return pair; });
+        std::sort(tData.begin(),tData.end(),[](const std::pair<T,size_t>& a, std::pair<T,size_t>& b){ return a.second < b.second;});
+        auto first = std::find_if(tData.cbegin(),tData.cend(),[start](const std::pair<T,size_t>& a){return a.first>=start;});
+        auto last = std::find_if(first,tData.cend(),[end](const std::pair<T,size_t>& a){return a.first>end;});
```
If, as I think you probably mean, the data to be sorted on the first field, then you can probably use lower_bound/upper_bound here, which performs a binary search, rather than the linear search that find_if uses.

highperformancecoder commented on libscidavis/src/Filter.h:
```diff
@@ -45,14 +45,14 @@ class Filter : public QObject
 public:
     Filter(ApplicationWindow *parent, Table *t = 0, QString name = QString());
     Filter(ApplicationWindow *parent, Graph *g = 0, QString name = QString());
-    ~Filter();
```
Maybe should make this virtual, rather than deleting it.

Suthiro replied:
Added an empty virtual destructor.

highperformancecoder commented on libscidavis/src/FFT.cpp:
```diff
@@ -72,47 +70,53 @@ void FFT::init()
+        amp.resize(d_x.size());
+        wavetable = gsl_fft_complex_wavetable_alloc(d_x.size());
+        workspace = gsl_fft_complex_workspace_alloc(d_x.size());
+        if ((nullptr == wavetable) || (nullptr == workspace))
+            throw std::bad_alloc();
+    } catch (const std::bad_alloc &) {
```
This is good, but will still leak memory if workspace allocation fails. Can be fixed by gsl_fft_complex_wavetable_free'ing the pointers in the catch block (hopefully it accepts nullptrs). Even better is an RAII design, where the _alloc is called in the constructor and _free in the destructor, as it makes the logic easier to follow, and automatically correct.

Suthiro replied:
GSL docs do not state anything about nullptrs passed to `free` functions, so I just decided to free workspace and wavetable with nullptr checks in catch block. About RAII design: `fftTable()` is called in the end, so I'm not sure if it is good to allocate everything in `FFT` constructor.

highperformancecoder commented:
I guess you've run out of steam now. The biggest issue is the index sort method in Filter.h, which looks incorrect  to me. There are a couple of other minor warts, which I would be happy to deal with in another PR.

Suthiro commented:
TBH I was waiting for feedback. Why do you think that index sorting is incorrect?

highperformancecoder reviewed (commented):
> TBH I was waiting for feedback. Why do you think that index sorting is incorrect?

You are sorting on an already sorted field. I left a comment around the code  concerned.

Suthiro committed:
Sorting bugfix & review-based improvements

Suthiro commented:
I don't know what is wrong with MSVC builds (again).
